File: src/main/java/main/data/Team.java
package main.data;

import com.google.common.collect.ImmutableList;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import javax.persistence.*;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;


@Entity
public class Team extends UserDetailsImpl implements Comparable<Team> {
    private int id;
    @NotBlank
    private String school;
    @Min(0)
    private int score;

    @Enumerated(EnumType.STRING)
    @NotNull
    private Division division;

    @NotNull
    @Size(max = 3)
    @OneToMany(
            mappedBy = "team",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private List<Individual> individuals;

    protected Team() {
    }

    public Team(int id, String password, String school, Division division) {
        super("team" + id, password);
        this.id = id;
        this.school = school;
        this.division = division;
        this.individuals = new ArrayList<>(3);

        for (int i = 0; i < 3; i++) {
            individuals.add(new Individual("", this, division));
        }
    }

    public int getId() {
        return id;
    }

    public String getSchool() {
        return school;
    }

    public void setSchool(String school) {
        this.school = school;
    }

    public int getScore() {
        return score;
    }

    public void setScore(int score) {
        this.score = score;
    }

    public List<Individual> getIndividuals() {
        return Collections.unmodifiableList(individuals);
    }

    public Division getDivision() {
        return division;
    }

    public void setDivision(Division division) {
        this.division = division;
        for (Individual i : individuals) {
            i.setDivision(division);
        }
    }

    @Override
    public int compareTo(Team o) {
        if (o.score != score) {
            return Integer.compare(o.score, score);
        }
        return Integer.compare(id, o.id);
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return ImmutableList.of(new SimpleGrantedAuthority("ROLE_TEAM"));
    }
}

File: src/main/kotlin/main/data/Team.kt
package main.data

import kotlinx.collections.immutable.persistentListOf
import kotlinx.collections.immutable.toImmutableList
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority

import javax.persistence.*
import java.util.ArrayList
import javax.validation.constraints.*

@Entity
class Team(val id: Int, password: String, @field:NotBlank var school: String, division: Division) : UserDetails("team$id", password), Comparable<Team> {
    @Enumerated(EnumType.STRING)
    @NotNull
    var division: Division = division
        set(division) {
            field = division
            for (i in individuals) {
                i.division = division
            }
        }

    @Min(0)
    @Max(720)
    var score: Int = 0

    @Min(0)
    var prestige: Int = 0

    @NotNull
    @Size(max = 3)
    @OneToMany(mappedBy = "team", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.EAGER)
    private val individuals: List<Individual>

    init {
        this.individuals = ArrayList(3)
        for (i in 0..2) {
            individuals.add(Individual("", this, division))
        }
    }

    fun getIndividuals(): List<Individual> =
            individuals.toImmutableList()

    override fun compareTo(other: Team): Int =
            if (other.score != score) {
                other.score.compareTo(score)
            } else id.compareTo(other.id)

    override fun getAuthorities(): Collection<GrantedAuthority> =
            persistentListOf(SimpleGrantedAuthority("ROLE_TEAM"))
}
